File: desktop/native/SurveyController.App/Services/BackendClient.cpp
#include "pch.h"
#include "BackendClient.h"
#include "JsonHelpers.h"

#include <condition_variable>
#include <filesystem>
#include <thread>

namespace winrt::SurveyController::App::Services
{
    namespace
    {
        constexpr std::uint32_t maxFrameSize = 16U << 20;

        [[noreturn]] void ThrowLastError(wchar_t const* operation)
        {
            throw winrt::hresult_error(HRESULT_FROM_WIN32(GetLastError()), operation);
        }
    }

    struct RpcSession
    {
        winrt::handle process;
        winrt::handle stdinWrite;
        winrt::handle stdoutRead;
        std::mutex ioMutex;
        std::mutex inFlightMutex;
        std::condition_variable inFlightChanged;
        std::size_t inFlight{};
        std::atomic_bool accepting{ true };
    };

    namespace
    {
        struct InFlightGuard
        {
            explicit InFlightGuard(std::shared_ptr<RpcSession> value) : session(std::move(value))
            {
                std::scoped_lock lock(session->inFlightMutex);
                if (!session->accepting.load())
                {
                    throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_SHUTDOWN_IN_PROGRESS), L"后端正在关闭");
                }
                ++session->inFlight;
            }

            ~InFlightGuard()
            {
                std::scoped_lock lock(session->inFlightMutex);
                --session->inFlight;
                session->inFlightChanged.notify_all();
            }

            std::shared_ptr<RpcSession> session;
        };
    }

    BackendClient& BackendClient::Current()
    {
        static BackendClient client;
        return client;
    }

    BackendClient::~BackendClient()
    {
        ShutdownImmediate();
    }

    BackendState BackendClient::State() const noexcept
    {
        std::scoped_lock lock(m_stateMutex);
        return m_state;
    }

    void BackendClient::Start()
    {
        {
            std::unique_lock lock(m_stateMutex);
            m_stateChanged.wait(lock, [&] { return m_state != BackendState::Starting; });
            if (m_state == BackendState::Running)
            {
                if (m_session && WaitForSingleObject(m_session->process.get(), 0) == WAIT_TIMEOUT) return;
                m_session.reset();
                m_state = BackendState::Stopped;
            }
            if (m_state == BackendState::Starting)
                throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_BUSY), L"后端正在启动");
            if (m_state == BackendState::Stopping)
                throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_SHUTDOWN_IN_PROGRESS), L"后端正在关闭");
            m_state = BackendState::Starting;
        }

        std::shared_ptr<RpcSession> session;
        try { session = CreateSession(); }
        catch (...)
        {
            std::scoped_lock lock(m_stateMutex);
            m_state = BackendState::Stopped;
            m_stateChanged.notify_all();
            throw;
        }

        std::scoped_lock lock(m_stateMutex);
        if (m_state != BackendState::Starting)
        {
            session->accepting.store(false);
            TerminateProcess(session->process.get(), ERROR_PROCESS_ABORTED);
            m_stateChanged.notify_all();
            throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_CANCELLED), L"后端启动已取消");
        }
        m_session = std::move(session);
        m_state = BackendState::Running;
        m_stateChanged.notify_all();
    }

    std::shared_ptr<RpcSession> BackendClient::AcquireSession()
    {
        for (;;)
        {
            {
                std::scoped_lock lock(m_stateMutex);
                if (m_state == BackendState::Running && m_session)
                {
                    if (WaitForSingleObject(m_session->process.get(), 0) == WAIT_TIMEOUT) return m_session;
                    m_session->accepting.store(false);
                    m_session.reset();
                    m_state = BackendState::Stopped;
                }
                if (m_state == BackendState::Stopping)
                    throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_SHUTDOWN_IN_PROGRESS), L"后端正在关闭");
            }
            Start();
        }
    }

    void BackendClient::MarkSessionFailed(std::shared_ptr<RpcSession> const& session) noexcept
    {
        session->accepting.store(false);
        if (session->process) TerminateProcess(session->process.get(), ERROR_PROCESS_ABORTED);
        std::scoped_lock lock(m_stateMutex);
        if (m_session == session && m_state == BackendState::Running)
        {
            m_session.reset();
            m_state = BackendState::Stopped;
            m_stateChanged.notify_all();
        }
    }

    void BackendClient::ShutdownImmediate() noexcept
    {
        std::shared_ptr<RpcSession> session;
        {
            std::scoped_lock lock(m_stateMutex);
            if (m_state == BackendState::Stopped) return;
            m_state = BackendState::Stopping;
            session = m_session;
            if (session) session->accepting.store(false);
        }

        if (session)
        {
            session->stdinWrite.close();
            if (session->process && WaitForSingleObject(session->process.get(), 0) == WAIT_TIMEOUT)
                TerminateProcess(session->process.get(), ERROR_PROCESS_ABORTED);

            std::unique_lock lock(session->inFlightMutex);
            session->inFlightChanged.wait_for(lock, std::chrono::seconds{ 2 }, [&] { return session->inFlight == 0; });
            lock.unlock();
            session->stdoutRead.close();
            session->process.close();
        }

        std::scoped_lock lock(m_stateMutex);
        m_session.reset();
        m_state = BackendState::Stopped;
        m_stateChanged.notify_all();
    }

    hstring BackendClient::Call(hstring const& method, hstring const& params, std::chrono::milliseconds timeout)
    {
        if (timeout <= std::chrono::milliseconds::zero())
            throw winrt::hresult_error(E_INVALIDARG, L"RPC 超时时间无效");

        auto session = AcquireSession();
        InFlightGuard inFlight{ session };
        std::scoped_lock ioLock(session->ioMutex);
        if (!session->accepting.load())
            throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_SHUTDOWN_IN_PROGRESS), L"后端正在关闭");

        auto const requestId = m_nextRequestId.fetch_add(1);
        auto const payload = BuildRequestPayload(requestId, method, params);
        auto const requestSize = static_cast<std::uint32_t>(payload.size());

        std::mutex timerMutex;
        std::condition_variable timerChanged;
        bool completed = false;
        std::atomic_bool timedOut = false;
        std::thread watchdog([session, timeout, &timerMutex, &timerChanged, &completed, &timedOut]
        {
            std::unique_lock lock(timerMutex);
            if (!timerChanged.wait_for(lock, timeout, [&] { return completed; }))
            {
                timedOut.store(true);
                session->accepting.store(false);
                if (session->process) TerminateProcess(session->process.get(), ERROR_TIMEOUT);
            }
        });
        auto finishWatchdog = [&]
        {
            { std::scoped_lock lock(timerMutex); completed = true; }
            timerChanged.notify_one();
            watchdog.join();
        };

        try
        {
            WriteExact(session->stdinWrite.get(), &requestSize, sizeof(requestSize));
            WriteExact(session->stdinWrite.get(), payload.data(), requestSize);
            std::uint32_t responseSize = 0;
            ReadExact(session->stdoutRead.get(), &responseSize, sizeof(responseSize));
            if (responseSize == 0 || responseSize > maxFrameSize)
                throw winrt::hresult_error(E_FAIL, L"RPC 响应大小无效");
            std::string responsePayload(responseSize, '\0');
            ReadExact(session->stdoutRead.get(), responsePayload.data(), responseSize);
            finishWatchdog();
            if (timedOut.load())
            {
                MarkSessionFailed(session);
                throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_TIMEOUT), L"后端响应超时");
            }
            return ParseResponsePayload(requestId, responsePayload);
        }
        catch (...)
        {
            finishWatchdog();
            MarkSessionFailed(session);
            if (timedOut.load())
                throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_TIMEOUT), L"后端响应超时");
            throw;
        }
    }

    Windows::Foundation::IAsyncOperation<hstring> BackendClient::CallAsync(
        hstring method, hstring params, std::chrono::milliseconds timeout)
    {
        co_await winrt::resume_background();
        co_return Call(method, params, timeout);
    }

    std::string BackendClient::BuildRequestPayload(std::uint64_t requestId, hstring const& method, hstring const& params)
    {
        if (method.empty()) throw winrt::hresult_error(E_INVALIDARG, L"RPC 方法不能为空");
        Windows::Data::Json::JsonObject request;
        request.Insert(L"id", Windows::Data::Json::JsonValue::CreateNumberValue(static_cast<double>(requestId)));
        request.Insert(L"method", Windows::Data::Json::JsonValue::CreateStringValue(method));
        try { request.Insert(L"params", Windows::Data::Json::JsonValue::Parse(params)); }
        catch (winrt::hresult_error const&) { throw winrt::hresult_error(E_INVALIDARG, L"RPC 参数不是有效 JSON"); }
        auto payload = winrt::to_string(request.Stringify());
        if (payload.empty() || payload.size() > maxFrameSize)
            throw winrt::hresult_error(E_INVALIDARG, L"RPC 请求大小无效");
        return payload;
    }

    hstring BackendClient::ParseResponsePayload(std::uint64_t requestId, std::string const& payload)
    {
        Windows::Data::Json::JsonObject response;
        hstring parseError;
        if (!TryParseJsonObject(winrt::to_hstring(payload), response, parseError))
            throw winrt::hresult_error(E_FAIL, L"RPC 响应不是有效 JSON：" + parseError);
        if (!response.HasKey(L"id") || response.GetNamedValue(L"id").ValueType() != Windows::Data::Json::JsonValueType::Number)
            throw winrt::hresult_error(E_FAIL, L"RPC 响应缺少有效 id");
        if (static_cast<std::uint64_t>(response.GetNamedNumber(L"id")) != requestId)
            throw winrt::hresult_error(E_FAIL, L"RPC 响应编号不匹配");
        auto const hasResult = response.HasKey(L"result");
        auto const hasError = response.HasKey(L"error");
        if (!hasResult && !hasError)
            throw winrt::hresult_error(E_FAIL, L"RPC 响应缺少 result 或 error");
        if (hasError)
        {
            auto value = response.GetNamedValue(L"error");
            if (value.ValueType() != Windows::Data::Json::JsonValueType::Null)
            {
                if (value.ValueType() != Windows::Data::Json::JsonValueType::Object)
                    throw winrt::hresult_error(E_FAIL, L"RPC error 字段类型无效");
                auto error = value.GetObject();
                if (!error.HasKey(L"message") || error.GetNamedValue(L"message").ValueType() != Windows::Data::Json::JsonValueType::String)
                    throw winrt::hresult_error(E_FAIL, L"RPC error 缺少有效 message");
                throw winrt::hresult_error(E_FAIL, error.GetNamedString(L"message"));
            }
        }
        if (!hasResult) throw winrt::hresult_error(E_FAIL, L"RPC 成功响应缺少 result");
        return response.GetNamedValue(L"result").Stringify();
    }

    std::shared_ptr<RpcSession> BackendClient::CreateSession()
    {
        auto session = std::make_shared<RpcSession>();
        SECURITY_ATTRIBUTES security{ sizeof(security), nullptr, TRUE };
        winrt::handle childStdinRead, childStdoutWrite, childStderr;
        if (!CreatePipe(childStdinRead.put(), session->stdinWrite.put(), &security, 0) ||
            !CreatePipe(session->stdoutRead.put(), childStdoutWrite.put(), &security, 0))
            ThrowLastError(L"创建后端管道失败");
        if (!SetHandleInformation(session->stdinWrite.get(), HANDLE_FLAG_INHERIT, 0) ||
            !SetHandleInformation(session->stdoutRead.get(), HANDLE_FLAG_INHERIT, 0))
            ThrowLastError(L"配置后端管道失败");
        childStderr.attach(CreateFileW(L"NUL", GENERIC_WRITE, FILE_SHARE_READ | FILE_SHARE_WRITE,
            &security, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, nullptr));
        if (!childStderr) ThrowLastError(L"打开后端错误输出失败");

        SIZE_T attributeBytes = 0;
        InitializeProcThreadAttributeList(nullptr, 1, 0, &attributeBytes);
        std::vector<std::byte> storage(attributeBytes);
        auto attributes = reinterpret_cast<PPROC_THREAD_ATTRIBUTE_LIST>(storage.data());
        if (!InitializeProcThreadAttributeList(attributes, 1, 0, &attributeBytes))
            ThrowLastError(L"初始化后端句柄列表失败");
        struct Guard { PPROC_THREAD_ATTRIBUTE_LIST value; ~Guard() { DeleteProcThreadAttributeList(value); } } guard{ attributes };
        std::array<HANDLE, 3> inherited{ childStdinRead.get(), childStdoutWrite.get(), childStderr.get() };
        if (!UpdateProcThreadAttribute(attributes, 0, PROC_THREAD_ATTRIBUTE_HANDLE_LIST,
            inherited.data(), sizeof(inherited), nullptr, nullptr))
            ThrowLastError(L"限制后端继承句柄失败");

        STARTUPINFOEXW startup{};
        startup.StartupInfo.cb = sizeof(startup);
        startup.StartupInfo.dwFlags = STARTF_USESTDHANDLES;
        startup.StartupInfo.hStdInput = childStdinRead.get();
        startup.StartupInfo.hStdOutput = childStdoutWrite.get();
        startup.StartupInfo.hStdError = childStderr.get();
        startup.lpAttributeList = attributes;
        auto path = BackendPath();
        auto commandLine = L"\"" + path + L"\"";
        PROCESS_INFORMATION process{};
        if (!CreateProcessW(path.c_str(), commandLine.data(), nullptr, nullptr, TRUE,
            EXTENDED_STARTUPINFO_PRESENT | CREATE_NO_WINDOW, nullptr, nullptr, &startup.StartupInfo, &process))
            ThrowLastError(L"启动 SurveyController 后端失败");
        CloseHandle(process.hThread);
        session->process.attach(process.hProcess);
        return session;
    }

    std::wstring BackendClient::BackendPath()
    {
        std::wstring modulePath(32768, L'\0');
        auto length = GetModuleFileNameW(nullptr, modulePath.data(), static_cast<DWORD>(modulePath.size()));
        if (length == 0 || length == modulePath.size()) ThrowLastError(L"读取应用路径失败");
        modulePath.resize(length);
        return (std::filesystem::path(modulePath).parent_path() / L"SurveyController.Backend.exe").wstring();
    }

    void BackendClient::ReadExact(HANDLE handle, void* buffer, std::uint32_t size)
    {
        auto bytes = static_cast<std::byte*>(buffer);
        for (std::uint32_t offset = 0; offset < size;)
        {
            DWORD read = 0;
            if (!ReadFile(handle, bytes + offset, size - offset, &read, nullptr)) ThrowLastError(L"读取后端响应失败");
            if (read == 0) throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_BROKEN_PIPE), L"后端响应提前结束");
            offset += read;
        }
    }

    void BackendClient::WriteExact(HANDLE handle, void const* buffer, std::uint32_t size)
    {
        auto bytes = static_cast<std::byte const*>(buffer);
        for (std::uint32_t offset = 0; offset < size;)
        {
            DWORD written = 0;
            if (!WriteFile(handle, bytes + offset, size - offset, &written, nullptr)) ThrowLastError(L"写入后端请求失败");
            if (written == 0) throw winrt::hresult_error(HRESULT_FROM_WIN32(ERROR_WRITE_FAULT), L"后端请求未写入");
            offset += written;
        }
    }
}

File: desktop/native-wpf/src/SurveyController.Core/Rpc/BackendClient.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace SurveyController.Core.Rpc;

/// <summary>
/// Go 后端（SurveyController.Backend.exe）的匿名管道 JSON-RPC 客户端。
/// 语义与 desktop/native 的 Services/BackendClient.cpp 保持一致：
/// 惰性启动、崩溃自动重启、按会话串行 I/O、超时看门狗终止进程。
/// </summary>
public sealed class BackendClient : IDisposable
{
    public static TimeSpan DefaultTimeout { get; } = TimeSpan.FromSeconds(15);

    public static BackendClient Current { get; } = new();

    private readonly object _stateGate = new();
    private readonly Func<BackendTransport> _transportFactory;
    private BackendState _state = BackendState.Stopped;
    private RpcSession? _session;
    private long _lastRequestId;

    public enum BackendState
    {
        Stopped,
        Starting,
        Running,
        Stopping,
    }

    public BackendClient()
        : this(() => ProcessBackendTransport.Start(ProcessBackendTransport.BackendPath()))
    {
    }

    /// <summary>测试可注入自定义传输工厂；生产路径拉起真实后端进程。</summary>
    public BackendClient(Func<BackendTransport> transportFactory)
    {
        _transportFactory = transportFactory;
    }

    public BackendState State
    {
        get
        {
            lock (_stateGate)
            {
                return _state;
            }
        }
    }

    /// <summary>同步执行一次 RPC 调用，返回 result 字段的字符串化 JSON。</summary>
    public string Call(string method, string paramsJson = "null", TimeSpan? timeout = null)
    {
        var effectiveTimeout = timeout ?? DefaultTimeout;
        if (effectiveTimeout <= TimeSpan.Zero)
        {
            throw new ArgumentOutOfRangeException(nameof(timeout), "RPC 超时时间无效");
        }

        var session = AcquireSession();
        lock (session.FlightGate)
        {
            if (!session.Accepting)
            {
                throw new RpcException("后端正在关闭");
            }
            session.InFlight++;
        }

        try
        {
            ulong requestId;
            string payload;
            string response;
            lock (session.IoGate)
            {
                if (!session.Accepting)
                {
                    throw new RpcException("后端正在关闭");
                }

                requestId = (ulong)Interlocked.Increment(ref _lastRequestId);
                payload = BuildRequestPayload(requestId, method, paramsJson);

                var completed = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
                var timedOut = false;
                var watchdog = Task.Run(async () =>
                {
                    var finished = await Task.WhenAny(completed.Task, Task.Delay(effectiveTimeout)).ConfigureAwait(false);
                    if (finished != completed.Task)
                    {
                        timedOut = true;
                        session.Accepting = false;
                        session.Transport.Kill();
                    }
                });

                try
                {
                    RpcFrame.WriteFrame(session.Transport.Stdin, payload);
                    response = RpcFrame.ReadFrame(session.Transport.Stdout);
                    completed.TrySetResult(true);
                    watchdog.Wait();
                    if (timedOut)
                    {
                        MarkSessionFailed(session);
                        throw new RpcTimeoutException("后端响应超时");
                    }
                    return ParseResponsePayload(requestId, response);
                }
                catch
                {
                    completed.TrySetResult(true);
                    watchdog.Wait();
                    MarkSessionFailed(session);
                    if (timedOut)
                    {
                        throw new RpcTimeoutException("后端响应超时");
                    }
                    throw;
                }
            }
        }
        finally
        {
            lock (session.FlightGate)
            {
                session.InFlight--;
                Monitor.PulseAll(session.FlightGate);
            }
        }
    }

    /// <summary>在线程池上执行 RPC 调用；调用方负责回到 UI 线程更新界面。</summary>
    public Task<string> CallAsync(string method, string paramsJson = "null", TimeSpan? timeout = null)
    {
        return Task.Run(() => Call(method, paramsJson, timeout));
    }

    public void Start()
    {
        lock (_stateGate)
        {
            while (_state == BackendState.Starting)
            {
                Monitor.Wait(_stateGate);
            }

            if (_state == BackendState.Running)
            {
                if (_session is { } running && running.Transport.IsAlive)
                {
                    return;
                }
                if (_session is { } dead)
                {
                    dead.Accepting = false;
                    _session = null;
                }
                _state = BackendState.Stopped;
            }

            if (_state == BackendState.Stopping)
            {
                throw new RpcException("后端正在关闭");
            }
            _state = BackendState.Starting;
        }

        RpcSession session;
        try
        {
            session = new RpcSession(_transportFactory());
        }
        catch
        {
            lock (_stateGate)
            {
                _state = BackendState.Stopped;
                Monitor.PulseAll(_stateGate);
            }
            throw;
        }

        lock (_stateGate)
        {
            if (_state != BackendState.Starting)
            {
                session.Accepting = false;
                session.Transport.Kill();
                Monitor.PulseAll(_stateGate);
                throw new RpcException("后端启动已取消");
            }
            _session = session;
            _state = BackendState.Running;
            Monitor.PulseAll(_stateGate);
        }
    }

    public void ShutdownImmediate()
    {
        RpcSession? session;
        lock (_stateGate)
        {
            if (_state == BackendState.Stopped)
            {
                return;
            }
            _state = BackendState.Stopping;
            session = _session;
            if (session is not null)
            {
                session.Accepting = false;
            }
        }

        if (session is not null)
        {
            try
            {
                session.Transport.Stdin.Dispose();
            }
            catch (Exception)
            {
            }
            session.Transport.Kill();

            lock (session.FlightGate)
            {
                if (session.InFlight > 0)
                {
                    Monitor.Wait(session.FlightGate, TimeSpan.FromSeconds(2));
                }
            }
            try
            {
                session.Transport.Stdout.Dispose();
            }
            catch (Exception)
            {
            }
            session.Transport.Dispose();
        }

        lock (_stateGate)
        {
            _session = null;
            _state = BackendState.Stopped;
            Monitor.PulseAll(_stateGate);
        }
    }

    public void Shutdown() => ShutdownImmediate();

    public void Dispose() => ShutdownImmediate();

    private RpcSession AcquireSession()
    {
        while (true)
        {
            lock (_stateGate)
            {
                if (_state == BackendState.Running && _session is { } current)
                {
                    if (current.Transport.IsAlive)
                    {
                        return current;
                    }
                    current.Accepting = false;
                    _session = null;
                    _state = BackendState.Stopped;
                }
                if (_state == BackendState.Stopping)
                {
                    throw new RpcException("后端正在关闭");
                }
            }
            Start();
        }
    }

    private void MarkSessionFailed(RpcSession session)
    {
        session.Accepting = false;
        session.Transport.Kill();
        lock (_stateGate)
        {
            if (ReferenceEquals(_session, session) && _state == BackendState.Running)
            {
                _session = null;
                _state = BackendState.Stopped;
                Monitor.PulseAll(_stateGate);
            }
        }
    }

    /// <summary>构造请求载荷 {"id":n,"method":"...","params":...}；params 必须是合法 JSON。</summary>
    public static string BuildRequestPayload(ulong requestId, string method, string paramsJson)
    {
        if (string.IsNullOrEmpty(method))
        {
            throw new ArgumentException("RPC 方法不能为空", nameof(method));
        }

        var request = new JsonObject
        {
            ["id"] = JsonValue.Create(requestId),
            ["method"] = method,
        };
        try
        {
            request["params"] = JsonNode.Parse(paramsJson ?? "null");
        }
        catch (JsonException exception)
        {
            throw new ArgumentException("RPC 参数不是有效 JSON", nameof(paramsJson), exception);
        }

        var payload = request.ToJsonString();
        if (payload.Length == 0 || Encoding.UTF8.GetByteCount(payload) > RpcFrame.MaxFrameSize)
        {
            throw new ArgumentException("RPC 请求大小无效", nameof(paramsJson));
        }
        return payload;
    }

    /// <summary>校验响应信封并返回 result 字段的字符串化 JSON。</summary>
    public static string ParseResponsePayload(ulong requestId, string payload)
    {
        JsonNode? root;
        try
        {
            root = JsonNode.Parse(payload);
        }
        catch (JsonException exception)
        {
            throw new RpcException($"RPC 响应不是有效 JSON：后端响应格式无效：{exception.Message}");
        }
        if (root is not JsonObject response)
        {
            throw new RpcException("RPC 响应不是有效 JSON：后端响应格式无效");
        }

        if (response["id"] is not JsonValue idValue || !idValue.TryGetValue<ulong>(out var responseId))
        {
            throw new RpcException("RPC 响应缺少有效 id");
        }
        if (responseId != requestId)
        {
            throw new RpcException("RPC 响应编号不匹配");
        }

        var hasResult = response.ContainsKey("result");
        var hasError = response.ContainsKey("error");
        if (!hasResult && !hasError)
        {
            throw new RpcException("RPC 响应缺少 result 或 error");
        }
        if (hasError && response["error"] is not null)
        {
            if (response["error"] is not JsonObject error)
            {
                throw new RpcException("RPC error 字段类型无效");
            }
            if (error["message"] is not JsonValue messageValue || !messageValue.TryGetValue<string>(out var message))
            {
                throw new RpcException("RPC error 缺少有效 message");
            }
            throw new RpcException(message);
        }
        if (!hasResult)
        {
            throw new RpcException("RPC 成功响应缺少 result");
        }
        // result 为 JSON null 时索引器返回 C# null，按字面量 "null" 返回。
        return response["result"]?.ToJsonString() ?? "null";
    }

    private sealed class RpcSession
    {
        public RpcSession(BackendTransport transport)
        {
            Transport = transport;
        }

        public BackendTransport Transport { get; }

        /// <summary>同一会话的管道 I/O 串行化锁，语义同 C++ ioMutex。</summary>
        public object IoGate { get; } = new();

        /// <summary>在途调用计数锁；关闭时等待在途调用排空。</summary>
        public object FlightGate { get; } = new();

        public int InFlight { get; set; }

        public bool Accepting { get; set; } = true;
    }
}
